### PR TITLE
Pull enabled for customizing tables

### PR DESCRIPTION
**Pull request by ThomasErgin**
See issue  [6136](https://github.com/abapGit/abapGit/issues/6136).

This commit enables you to download table-content to your system.
For safety reasons, this is limited to customizing tables (DD09L-TABART = 'APPL2') in custom tables (DD09L-AS4USER <> 'SAP'). Only those tables will be written to the system, that has been selected by the user. If the system has been set to change recording in SCC4, the user will be prompted for a customizing request, where the entries will be added to.

What is missing:
> The solution does not fit perfectly in the existing architecture. I am thinking of moving the whole persistance logic to another place outside the deserializer.
> Update of existing table entries is not working, only INSERT/DELETE is allowed.

### CODE REVIEW

File: src/data/zcl_abapgit_data_utils.clas.abap [hunk 1]
```diff
@@ -71,7 +72,7 @@ CLASS zcl_abapgit_data_utils IMPLEMENTATION.
         CALL METHOD lo_obj->('IF_XCO_DBT_FIELDS~GET_NAMES')
           RECEIVING
             rt_names = rt_names.
-      CATCH cx_sy_dyn_call_illegal_class.
+      CATCH cx_sy_dyn_call_illegal_class cx_xco_news_exception.
```
Comment (larshp) [inline src/data/zcl_abapgit_data_utils.clas.abap:75]: `cx_xco_news_exception` is not released, perhaps use `cx_xco_runtime_exception` instead?

https://abapedia.org/steampunk-2302-api/

Comment (larshp) [reply src/data/zcl_abapgit_data_utils.clas.abap:75]: and `cx_xco_runtime_exception` probably does not exists in 702, so it cannot be used either

and `cx_xco_runtime_exception` is a `cx_no_check`

what kind of error does the current code create in the system?

Comment (ThomasErgin) [reply src/data/zcl_abapgit_data_utils.clas.abap:75]: When using the feature "Beta - Data" to add a table to my repository, the mentioned exception will be thrown on my machine and the table wont be added.

**Comment by larshp** [comment]
@ThomasErgin please set the status of this PR to draft, https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

mark it as ready when you've finished doing changes

File: src/data/zcl_abapgit_data_utils.clas.abap (hunk 1 above)
Comment (larshp) [reply src/data/zcl_abapgit_data_utils.clas.abap]: the abapGit code must be 702 compatible

`cx_xco_runtime_exception` does not even exist in 750

**Review by larshp — CHANGES_REQUESTED** [review]
check results from static analysis https://github.com/abapGit/abapGit/pull/6138/checks?check_run_id=11884406184

the code must be 702 compatible

**Comment by mbtools** [comment]
Hi @ThomasErgin 

our dev guidelines and checks can be intimidating at the beginning (https://docs.abapgit.org/development/guidelines.html). I put this into draft for now and am happy to help if you get stuck.

As Lars wrote, you can't use 740 syntax like inline data declarations or array reads. Also all variables and parameters must have proper prefixes. It's can take a few rounds of fixing, getting new lint errors, and fixing but it's worth the effort to keep our code consistent. 😃 

PS: I didn't know about `CL_TABLE_UTILITIES_BRF`. Nice!

**Comment by mbtools** [comment]
And addition check could calling function `CHECK_CUSTOMER_NAMES` for each record to be changed.

**Comment by larshp** [comment]
`IF NOT is_table_allowed_to_edit( iv_table_name = lv_table_name is_checks = is_checks ).` is not valid 702 syntax

an IF in 702 must be

`IF <something> <compare> <somethingElse>.` 

eg. 
`IF is_table_allowed_to_edit( iv_table_name = lv_table_name is_checks = is_checks ) = abap_false.`

is valid 702

File: src/data/zcl_abapgit_data_utils.clas.abap [hunk 2]
```diff
@@ -71,7 +72,7 @@ CLASS zcl_abapgit_data_utils IMPLEMENTATION.
         CALL METHOD lo_obj->('IF_XCO_DBT_FIELDS~GET_NAMES')
           RECEIVING
             rt_names = rt_names.
-      CATCH cx_sy_dyn_call_illegal_class.
+      CATCH cx_sy_dyn_call_illegal_class cx_xco_news_exception. "XCO exception might be thrown in newer SAP releases
```
Comment (larshp) [inline src/data/zcl_abapgit_data_utils.clas.abap:75]: `cx_xco_news_exception` does not exist in 702

**Comment by larshp** [comment]
changes to `zcl_abapgit_data_serializer.clas.abap` was made in https://github.com/abapGit/abapGit/pull/6157, dont revert it with this PR

no changes expected to `.abapgit.xml`

File: src/data/zcl_abapgit_data_utils.clas.abap [hunk 3]
```diff
@@ -71,7 +72,7 @@
         CALL METHOD lo_obj->('IF_XCO_DBT_FIELDS~GET_NAMES')
           RECEIVING
             rt_names = rt_names.
-      CATCH cx_sy_dyn_call_illegal_class.
+        CATCH cx_sy_dyn_call_illegal_class cx_xco_runtime_exception. "XCO exception might be thrown in newer SAP releases
```
[the remark this answers is not in the document]
Comment (mbtools) [reply src/data/zcl_abapgit_data_utils.clas.abap:75]: suggest `catch cx_root.`

**Comment by ThomasErgin** [comment]
Hi @mbtools,
I addressed your remarks 1 and 3. Instead of reverting the changes `zcl_abapgit_data_utils` I am catching now `CX_NO_CHECK` (parent of `CX_XCO_RUNTIME_EXCEPTION`) which should exist in old SAP releases uch as 702. I hope this is fine for you. 

Regarding your remark on `as4user <> 'SAP'`: Of cause, this will prevent adding standard customizing to a repository. When I decided to add this, I considered how I am using ABAPGIT right now. I seldomly feel the need of moving SAP standard customizing around (this I will usually do with TMS). I use it more for sharing tools with other SAP users and would like to include the customizing of those tools - usually sitting in custom tables. 
If ABAPGIT would allow to pull standard customizing, a repository might accidentally overwrite my local systems settings, which I see as a risk.
Cheers 
Thomas

**Review by mbtools — CHANGES_REQUESTED** [review]
there are more scenarios than just custom tables. in any case, it's the responsibility of the developer to check and decide what changes should be pulled into a system.

it's a start and we can change or enhance the check later.

cx_no_check will work with 702 so we can merge it this way. just fix the abapgit.xml, please

**Comment by mbtools** [comment]
i believe yes but will check again. i'm testing it and got a couple dumps. i will update the PR in a bit

**Comment by mbtools** [comment]
Updated test repo: https://github.com/abapGit-tests/data_custom_table

Dumps:

image

image

image

image

- Changed table exists check to `DD_EXISTS_TABLE` and move to `zcl_abapgit_data_utils`
- Added exists check to `build_table_itab` wgich fixes `NOT_FOUND` exception in `zcl_abapgit_data_utils`
- Fixed "generic reference cannot be dereferenced" syntax error in `zcl_abapgit_data_deserializer`

**Comment by mbtools** [comment]
Confirmed: `TREDT_OBJECTS` & `CL_TABLE_UTILITIES_BRF` exist in 702

File: src/data/zcl_abapgit_data_deserializer.clas.abap [hunk 4]
```diff
@@ -73,6 +73,33 @@ CLASS zcl_abapgit_data_deserializer IMPLEMENTATION.
...
+  METHOD is_table_allowed_to_edit.
+
+    DATA lv_tabname TYPE dd02l-tabname.
+
+    "Did the user flagged this table for update?
+    READ TABLE is_checks-overwrite TRANSPORTING NO FIELDS
+      WITH KEY obj_name = iv_table_name decision = zif_abapgit_definitions=>c_yes.
+    IF sy-subrc <>  0.
+      RETURN.
+    ENDIF.
```
Comment (larshp) [inline src/data/zcl_abapgit_data_deserializer.clas.abap:85]: this check is related to the TABL object? ie. not the data, ie. the data cannot be updated unless the table is also changed?

File: src/data/zcl_abapgit_data_deserializer.clas.abap [hunk 5]
```diff
@@ -73,6 +73,33 @@ CLASS zcl_abapgit_data_deserializer IMPLEMENTATION.
...
+        ON dd09l~tabname = dd02l~tabname
+        AND dd09l~as4local = dd02l~as4local
+        AND dd09l~as4vers = dd02l~as4vers
+      INTO lv_tabname
+      WHERE dd09l~tabname = iv_table_name
+        AND dd09l~tabart = 'APPL2'
+        AND dd09l~as4user <> 'SAP'
+        AND dd09l~as4local = 'A' "Only active tables
+        AND dd02l~contflag = 'C'. "Only customizing tables
+    rv_allowed_to_edit = boolc( sy-subrc = 0 ).
```
Comment (larshp) [inline src/data/zcl_abapgit_data_deserializer.clas.abap:99]: I need a user exit here, or some of the stuff I'm working on will not work anymore

Comment (larshp) [reply src/data/zcl_abapgit_data_deserializer.clas.abap:99]: eg. see the breaking cross check in https://github.com/abapGit/abapGit/pull/6138/checks?check_run_id=12341891700

**Comment by larshp** [comment]
`DD_EXISTS_TABLE`: no, its not released, the previous code works on both on-prem + steampunk + open-abap

**Comment by larshp** [comment]
also see https://github.com/abapGit/abapGit/issues/6136#issuecomment-1461720552

> I think showing a full overview of the consequences/actions to the user, before confirming the changes might solve it

pulling the data without showing the information first to the user I think is a problem. As noted, suggest showing a full overview, and then having the user confirm that. Ie. separate R3TR and data contents into two steps

File: src/data/zcl_abapgit_data_deserializer.clas.abap (hunk 4 above)
Comment (mbtools) [reply src/data/zcl_abapgit_data_deserializer.clas.abap:85]: there's a separate checkbox for TABU but you are right that the read must include `obj_type = 'TABU'`

File: src/data/zcl_abapgit_data_deserializer.clas.abap (hunk 5 above)
Comment (mbtools) [reply src/data/zcl_abapgit_data_deserializer.clas.abap:99]: agree. i will add an exit

**Comment by mbtools** [comment]
- [ ] Remove `DD_EXISTS_TABLE` (and investigate scenario that dumped)
- [ ] Add exit
- [ ] Add popup showing data changes (separate PR)

Sorry, @ThomasErgin, as you can see it's more complex. I will make the necessary changes so we get something that works for all.

**Comment by larshp** [comment]
please do a page instead of a popup, a page can also give a better structured overview. Tho this is something that can be done in a follow up PR, it will have to refactor the structural changes done in this PR

**Comment by mbtools** [comment]
Alright, this is looking much better and working well:

- By default, only customer-defined customizing tables are supported for writing data
- New exits `change_supported_data_objects` can be used to add other objects to the supported list (see example below)
- New class `zcl_abapgit_data_supporter` is used to check if object is supported (included factory and injector for testing)
  *If you have a better name for the class, let me know and I will rename it*
- Table entries are only added to transport for customizing tables. Non-customizing tables added via exit do not create transport entries
  *Should we add those to a workbench transport? (I would say, no)*
- Currently, user is prompted for changes to `TABU <table>` in the regular pull popup. More elaborate display of changes and confirmation on a separate page will be a new PR. 
- Test repos:
  - https://github.com/abapGit-tests/TABU_custom_table
  - https://github.com/abapGit-tests/TABU_sap_table (requires adding `RSADMIN` to exit)

Exit example:

image

As you see in the example, it's possible to use wildcards for adding tables.

**Comment by mbtools** [comment]
@larshp not sure what broke in https://github.com/open-abap/load-table-contents/

File: src/data/zcl_abapgit_data_deserializer.clas.testclasses.abap [hunk 6]
```diff
@@ -26,7 +26,6 @@ CLASS ltcl_test IMPLEMENTATION.
 
     li_config->add_config( ls_config ).
 
-* this does not change the database,
     li_cut->deserialize(
```
Comment (larshp) [inline src/data/zcl_abapgit_data_deserializer.clas.testclasses.abap:29]: will this unit test now do database changes?

Comment (mbtools) [reply src/data/zcl_abapgit_data_deserializer.clas.testclasses.abap:29]: it certainly shouldn't. only `actualize` updates DB

Comment (mbtools) [reply src/data/zcl_abapgit_data_deserializer.clas.testclasses.abap:30]: updated comment

**Comment by larshp** [comment]
https://github.com/open-abap/load-table-contents/ I'll fix that later

**Review by larshp — APPROVED** [review]

**Comment by mbtools** [comment]
Thanks @ThomasErgin for getting the ball rolling! 

Follow-ups: #6181, #6182

**Comment by larshp** [comment]
@ThomasErgin welcome to abapGit as a contributor 🎉

**Pull request merged**